### ISSUE TITLE
Clarification on Equation 5 in the Paper

### ISSUE THREAD

**Issue by zero2vec** [Dstate/LBP#4]
Hello, and thank you for sharing this impressive work.

I have few questions regarding Equation 5 in your paper.
According to the equation, the loss is computed by summing over all sub-goal scales $$i = 1 \dots n$$ for each time step $$t$$.

1. Does this mean that even when the condition $$t < \lambda^{i} < \lambda^{i-1} $$ holds, the model is still trained to predict $$z_{\lambda^{i}}$$ at step $$t$$?

2. Are the subgoal timesteps randomly sampled according to the distribution $$p(\{z_{\lambda^i}\}_{i=1}^{n} \mid \tau)$$ or is there another sampling strategy used in practice?

Any clarification would be greatly appreciated. Thank you again for your time and this excellent contribution!

**Comment by Dstate** [comment]
Hi! Thanks for your attention to our work!

1. In our paper, we denote each future subgoal as $z_{\lceil t + \lambda^{i}(H - t) \rceil}$, which we simplify as $z_{\lambda^{i}}$ for convenience.
For example, when $\lambda$ is set to 0.5, the backward planning model will generate the following subgoals in order: 
$z_{\lambda^{1}} = z_{\lceil t + \frac{1}{2}(H - t) \rceil}$
$z_{\lambda^{2}} = z_{\lceil t + \frac{1}{4}(H - t) \rceil}$
$z_{\lambda^{3}} = z_{\lceil t + \frac{1}{8}(H - t) \rceil}$
...
That said, $\lambda^{i} < \lambda^{i-1}$ always holds, and the corresponding indices $t < \lceil t + \lambda^{i}(H - t) \rceil < \lceil t + \lambda^{i-1}(H - t) \rceil$. 
The model first predicts the distant subgoal ($z_{\lambda^{i-1}}$), then the closer one ($z_{\lambda^{i}}$).

2. At each training step, we uniformly sample $t \in \lbrace 1, 2, \ldots, H \rbrace$ and $i \in \lbrace 1, 2, \ldots, n \rbrace$. Based on the sampled t and i, we can obtain corresponding $z_{\lambda ^ {i}} = z_{\lceil t + \lambda^{i}(H - t) \rceil}$ as prediction target. In practice, since $n$ is not large, we use all $n$ subgoals at each step to accelerate training.

If you have any other questions, please feel free to ask. I would be happy to discuss them further.

**Comment by zero2vec** [comment]
Thanks again for your clarification!
It seems I now have a much clearer understanding of how LBP works.

I have one more question!

In the comparison experiment between the backward and forward paradigms in the paper, LBP predicts 3 (sub)goals in total, whereas the forward paradigm predicts 10 subgoals at each timestep $t$.

I'm wondering — if the forward paradigm were restricted to predict only around 3 subgoals (similar to LBP), would the mean squared error (MSE) to the ground-truth subgoals in latent space be significantly reduced?

Has there been any analysis or experiment on this? Thank you!

**Comment by Dstate** [comment]
Yes, the compounding errors are closely related to the prediction times. If the forward paradigm were restricted to predict few subgoals, the compounding errors will be significantly reduced. However, as the forward planning typically employs a fixed planning horizon, (e.g., SuSie predicts the future image 15 steps ahead),  it requires multiple steps to reach distant goals. This makes it challenging to provide long-horizon guidance with high accuracy.
In contrast, LBP strikes a better balance between long-horizon guidance and prediction accuracy through its recursive planning strategy. Moreover, by generating subgoals under the constraint of the final goal, LBP naturally produces trajectories that remain well-aligned with the task objective.
